password/src/Data/Password/Validate.hs

```diff
@@ -126,137 +223,136 @@ data CharacterCategory
   -- ^ ASCII digits
   deriving (Eq, Ord, Show)
 
--- | Convert 'CharacterCategory' into associated predicate function
+-- | Convert a 'CharacterCategory' into its associated predicate function
+--
+-- @since 2.1.0.0
 categoryToPredicate :: CharacterCategory -> (Char -> Bool)
 categoryToPredicate = \case
   Uppercase -> isAsciiUpper
   Lowercase -> isAsciiLower
   Special -> isSpecial
   Digit -> isDigit
 
--- | Possible reason of 'Password' being invalid
+type MinimumLength = Int
+type MaximumLength = Int
+type ProvidedLength = Int
+type MinimumAmount = Int
+type ProvidedAmount = Int
+
+-- | Possible reasons for a 'Password' to be invalid.
+--
+-- @since 2.1.0.0
 data InvalidReason
-  = PasswordTooShort !Int !Int
+  = PasswordTooShort !MinimumLength !ProvidedLength
   -- ^ Length of 'Password' is too short.
-  --
-  -- Expected at least 'Int' characters but the actual length is 'Int'
-  | PasswordTooLong !Int !Int
+  | PasswordTooLong !MaximumLength !ProvidedLength
   -- ^ Length of 'Password' is too long.
-  --
-  -- Expected at maximum of 'Int' characters, but the actual length is 'Int'
-  | NotEnoughReqChars CharacterCategory !Int !Int
+  | NotEnoughReqChars !CharacterCategory !MinimumAmount !ProvidedAmount
   -- ^ 'Password' does not contain required number of characters.
-  --
-  -- Expected at least 'Int' characters of 'CharacterCategory' but the password only
-  -- contains 'Int'
   | InvalidCharacters !Text
   -- ^ 'Password' contains characters that cannot be used
-  | InvalidPasswordPolicy [InvalidPolicyReason]
-  -- ^ 'PasswordPolicy' is invalid
   deriving (Eq, Ord, Show)
 
--- | Possible reason of 'PasswordPolicy' being invalid
+-- | Possible reasons for a 'PasswordPolicy' to be invalid
+--
+-- @since 2.1.0.0
 data InvalidPolicyReason
-  = InvalidLength !Int !Int
+  = InvalidLength !ProvidedLength !ProvidedLength
```

Comment:
```suggestion
  = InvalidLength !MinimumLength !MaximumLength
```

Reply:
After reviewing this again, you actually had a point. In the `InvalidReason` type the `MinimumLength` and `MaximumLength` type synonyms are used to show what the policy stated, and the `ProvidedLength` is what the length of the to-be-validated password actually was.
I think it'd make more sense to then use `Min-/MaximumLength` in the `InvalidPolicyReason` type to keep it consistent and also more obvious.